db/db_impl_write.cc

```diff
@@ -994,8 +996,12 @@ Status DBImpl::WriteRecoverableState() {
       for (uint64_t sub_batch_seq = seq + 1;
            sub_batch_seq < next_seq && status.ok(); sub_batch_seq++) {
         uint64_t const no_log_num = 0;
+        // Unlock it since the callback might end up locking mutex. e.g.,
+        // AddCommitted -> AdvanceMaxEvictedSeq -> GetSnapshotListFromDB
+        mutex_.Unlock();
```

Comment:
I'm not sure if it's really safe just to unlock/relock db mutex like this, but I don't know the code here that well. Maybe someone from the rocksdb team can have a look.

Is there a possibility of another deadlock because we're now changing the lock acquisition order? I also don't know if there's code that assumes the lock was held continuously before and after `WriteRecoverableState`, and that we're breaking atomicity somehow.

One possible way to avoid this is to pass down the fact that we're already holding the db mutex, but that might be a lot of code.

Reply:
This is called at the head of the write queue where all the concurrent writes to db are stopped (https://github.com/facebook/rocksdb/blob/189e711b3740ae3fbe2eeb8cd5a12419346bd627/db/db_impl_write.cc#L1483). Releasing db_mutex_  and reacquiring it is thus safe and a common practice. Example: https://github.com/facebook/rocksdb/blob/189e711b3740ae3fbe2eeb8cd5a12419346bd627/db/db_impl_write.cc#L1503

Passing down the bool that db_mutex would make the API complicated.